lib/core/source/common/hive_base_source.dart

```diff
@@ -0,0 +1,99 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:dartx/dartx.dart';
+import 'package:flutter_template/core/source/common/local_storage.dart';
+import 'package:hive/hive.dart';
+import 'package:rxdart/rxdart.dart';
+
+abstract class HiveBaseSource<Model, Key> implements LocalStorage<Model, Key> {
+  final Map<String, dynamic> Function(Model) dbParser;
+  final Model Function(Map<String, dynamic>) modelParser;
+  Box<String>? _box;
+
+  HiveBaseSource({required this.dbParser, required this.modelParser});
+
+  Future<Box<String>> getBox() async {
+    if (_box != null && _box!.isOpen) {
+      return _box!;
+    } else {
+      _box = await Hive.openBox<String>(Model.toString());
+      return _box!;
+    }
+  }
+
+  Future<T> withBox<T>(Future<T> Function(Box<String>) body) async =>
+      body(await getBox());
+
+  @override
+  Future<Model> putElement(
+    Key key,
+    Model response,
+  ) async =>
+      withBox((box) async {
+        await box.put(
+          key,
+          jsonEncode(dbParser(response)),
+        );
+        return response;
+      });
+
+  @override
+  Future<List<Model>> putAllElements(Map<Key, Model> entries) =>
+      withBox((box) async {
+        await box.putAll(
+          entries.mapValues(
+            (entry) => jsonEncode(dbParser(entry.value)),
+          ),
+        );
+        return getElements();
+      });
+
+  @override
+  Future<void> deleteElement(Key key) => withBox((box) => box.delete(key));
+
+  @override
+  Future<void> deleteAllElements() => withBox((box) => box.clear());
+
+  @override
+  Future<Model?> getElement(
+    Key key,
+  ) =>
+      withBox((box) async {
+        final data = box.get(key);
+        return data == null
+            ? null
+            : modelParser(
+                jsonDecode(
+                  data,
+                ),
```

Comment:
Maybe the same line? or do you think it's worst ?
```suggestion
            : modelParser(jsonDecode(data),
```